src/sys/fanotify.rs

```diff
@@ -0,0 +1,365 @@
+    /// from the fanotify group, this event is to notify that a group queue
+    /// overflow occured.
+    pub fd: Option<OwnedFd>,
+    /// PID of the process that caused the event. TID in case flag
+    /// `FAN_REPORT_TID` was set at group initialization.
+    pub pid: i32,
+}
+
+impl FanotifyEvent {
+    /// Checks that compile fanotify API version is equal to the version of the
+    /// event.
+    pub fn has_compile_version(&self) -> bool {
+        self.version == FANOTIFY_METADATA_VERSION
+    }
+}
+
+#[derive(Debug)]
+/// Abstraction over the structure to be sent to allow or deny a given event.
+pub struct FanotifyResponse<'e> {
+    /// A borrow of the file descriptor from the structure `FanotifyEvent`.
+    pub fd: BorrowedFd<'e>,
+    /// Indication whether or not the permission is to be granted.
+    pub response: Response,
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+/// Response to be wrapped in `FanotifyResponse` and sent to the `Fanotify`
+/// group to allow or deny an event.
+pub enum Response {
+    /// Allow the event.
+    Allow,
+    /// Deny the event.
+    Deny,
+}
+
+/// A fanotify group. This is also a file descriptor that can feed to other
+/// interfaces consuming file descriptors.
+#[derive(Debug)]
+pub struct Fanotify {
+    fd: OwnedFd,
+}
+
+impl Fanotify {
+    /// Initialize a new fanotify group.
+    ///
+    /// Returns a Result containing a Fanotify instance.
+    ///
+    /// For more information, see [fanotify_init(2)](https://man7.org/linux/man-pages/man7/fanotify_init.2.html).
+    pub fn init(flags: InitFlags, event_f_flags: OFlags) -> Result<Fanotify> {
+        let res = Errno::result(unsafe {
+            libc::fanotify_init(flags.bits(), event_f_flags.bits())
+        });
+        res.map(|fd| Fanotify { fd: unsafe { OwnedFd::from_raw_fd(fd) }})
+    }
+
+    /// Add, remove, or modify an fanotify mark on a filesystem object.
+    /// If `dirfd` is `None`, `AT_FDCWD` is used.
+    ///
+    /// Returns a Result containing either `()` on success or errno otherwise.
+    ///
+    /// For more information, see [fanotify_mark(2)](https://man7.org/linux/man-pages/man7/fanotify_mark.2.html).
+    pub fn mark<P: ?Sized + NixPath>(
+        &self,
+        flags: MarkFlags,
+        mask: MaskFlags,
+        dirfd: Option<RawFd>,
+        path: Option<&P>,
+    ) -> Result<()> {
+        fn with_opt_nix_path<P, T, F>(p: Option<&P>, f: F) -> Result<T>
+        where
+            P: ?Sized + NixPath,
+            F: FnOnce(*const libc::c_char) -> T,
+        {
+            match p {
+                Some(path) => path.with_nix_path(|p_str| f(p_str.as_ptr())),
+                None => Ok(f(std::ptr::null())),
+            }
+        }
+
+        let res = with_opt_nix_path(path, |p| unsafe {
+            libc::fanotify_mark(
+                self.fd.as_raw_fd(),
+                flags.bits(),
+                mask.bits(),
+                dirfd.unwrap_or(libc::AT_FDCWD),
```

Comment:
We have a `at_rawfd()` function, which is used by all the `xxx_at(dirfd: Option<RawFd>, ..)` functions, it might be better to use it here to keep consistency, though they are actually the same thing

That function is gated with:

```
#[cfg(any(
    all(feature = "fs", not(target_os = "redox")),
    all(feature = "process", any(target_os = "android", target_os = "linux"))
))]
```

you need to add the `fanotify` feature to it